File: examples/ts/schnorr-zkp/SchnorrZKP.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { createHash } from 'node:crypto';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract } from 'runar-testing';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'SchnorrZKP.runar.ts'), 'utf8');

// secp256k1 constants
const EC_P = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2Fn;
const EC_N = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141n;
const GX = 0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798n;
const GY = 0x483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8n;

// ---------------------------------------------------------------------------
// JS EC helpers for test vector generation
// ---------------------------------------------------------------------------

function mod(a: bigint, m: bigint): bigint { return ((a % m) + m) % m; }

function modInv(a: bigint, m: bigint): bigint {
  let [old_r, r] = [mod(a, m), m];
  let [old_s, s] = [1n, 0n];
  while (r !== 0n) {
    const q = old_r / r;
    [old_r, r] = [r, old_r - q * r];
    [old_s, s] = [s, old_s - q * s];
  }
  return mod(old_s, m);
}

function pointAdd(x1: bigint, y1: bigint, x2: bigint, y2: bigint): [bigint, bigint] {
  if (x1 === x2 && y1 === y2) {
    const slope = mod(3n * x1 * x1 * modInv(2n * y1, EC_P), EC_P);
    const rx = mod(slope * slope - 2n * x1, EC_P);
    return [rx, mod(slope * (x1 - rx) - y1, EC_P)];
  }
  const slope = mod((y2 - y1) * modInv(x2 - x1, EC_P), EC_P);
  const rx = mod(slope * slope - x1 - x2, EC_P);
  return [rx, mod(slope * (x1 - rx) - y1, EC_P)];
}

function scalarMul(bx: bigint, by: bigint, k: bigint): [bigint, bigint] {
  k = mod(k, EC_N);
  let rx = bx, ry = by, started = false;
  for (let i = 255; i >= 0; i--) {
    if (started) [rx, ry] = pointAdd(rx, ry, rx, ry);
    if ((k >> BigInt(i)) & 1n) {
      if (!started) { rx = bx; ry = by; started = true; }
      else [rx, ry] = pointAdd(rx, ry, bx, by);
    }
  }
  return [rx, ry];
}

function bigintToHex32(n: bigint): string {
  return n.toString(16).padStart(64, '0').toUpperCase();
}

function makePointHex(x: bigint, y: bigint): string {
  return bigintToHex32(x) + bigintToHex32(y);
}

// ---------------------------------------------------------------------------
// Fiat-Shamir helpers
// ---------------------------------------------------------------------------

function sha256(hex: string): string {
  return createHash('sha256').update(Buffer.from(hex, 'hex')).digest('hex').toUpperCase();
}

function hash256Hex(hex: string): string {
  return sha256(sha256(hex));
}

/** bin2num: interpret hex bytes as little-endian signed integer (Bitcoin script number). */
function bin2num(hex: string): bigint {
  const bytes = Buffer.from(hex, 'hex');
  if (bytes.length === 0) return 0n;
  const negative = (bytes[bytes.length - 1] & 0x80) !== 0;
  const last = bytes[bytes.length - 1] & 0x7f;
  let result = BigInt(last);
  for (let i = bytes.length - 2; i >= 0; i--) {
    result = (result << 8n) | BigInt(bytes[i]);
  }
  return negative ? -result : result;
}

/** Derive Fiat-Shamir challenge: e = bin2num(hash256(R || P)) */
function deriveChallenge(rHex: string, pubKeyHex: string): bigint {
  return bin2num(hash256Hex(rHex + pubKeyHex));
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('SchnorrZKP contract', () => {
  it('verifies a valid Schnorr ZKP proof with Fiat-Shamir challenge', () => {
    const privKey = 42n;
    const [pubX, pubY] = scalarMul(GX, GY, privKey);
    const pubKeyHex = makePointHex(pubX, pubY);

    const r = 12345n;
    const [rX, rY] = scalarMul(GX, GY, r);
    const rHex = makePointHex(rX, rY);

    // Challenge derived via Fiat-Shamir: e = bin2num(hash256(R || P))
    const e = deriveChallenge(rHex, pubKeyHex);
    const s = mod(r + e * privKey, EC_N);

    const c = TestContract.fromSource(source, { pubKey: pubKeyHex });
    const result = c.call('verify', { rPoint: rHex, s });
    expect(result.success).toBe(true);
  });

  it('rejects a proof with wrong s value', () => {
    const privKey = 42n;
    const [pubX, pubY] = scalarMul(GX, GY, privKey);
    const pubKeyHex = makePointHex(pubX, pubY);

    const r = 12345n;
    const [rX, rY] = scalarMul(GX, GY, r);
    const rHex = makePointHex(rX, rY);

    const e = deriveChallenge(rHex, pubKeyHex);
    const s = mod(r + e * privKey, EC_N);

    const c = TestContract.fromSource(source, { pubKey: pubKeyHex });
    const result = c.call('verify', { rPoint: rHex, s: s + 1n });
    expect(result.success).toBe(false);
  });

  it('rejects a proof with wrong R point (tampered commitment)', () => {
    const privKey = 42n;
    const [pubX, pubY] = scalarMul(GX, GY, privKey);
    const pubKeyHex = makePointHex(pubX, pubY);

    // Compute valid proof for one R
    const r = 12345n;
    const [rX, rY] = scalarMul(GX, GY, r);
    const rHex = makePointHex(rX, rY);
    const e = deriveChallenge(rHex, pubKeyHex);
    const s = mod(r + e * privKey, EC_N);

    // Use a different R — the on-chain challenge will differ, breaking the proof
    const [rX2, rY2] = scalarMul(GX, GY, 99999n);
    const rHex2 = makePointHex(rX2, rY2);

    const c = TestContract.fromSource(source, { pubKey: pubKeyHex });
    const result = c.call('verify', { rPoint: rHex2, s });
    expect(result.success).toBe(false);
  });

  it('works with larger private key', () => {
    const privKey = 0xDEADBEEFCAFEn;
    const [pubX, pubY] = scalarMul(GX, GY, privKey);
    const pubKeyHex = makePointHex(pubX, pubY);

    const r = 0xABCDEF0123456789n;
    const [rX, rY] = scalarMul(GX, GY, r);
    const rHex = makePointHex(rX, rY);

    const e = deriveChallenge(rHex, pubKeyHex);
    const s = mod(r + e * privKey, EC_N);

    const c = TestContract.fromSource(source, { pubKey: pubKeyHex });
    const result = c.call('verify', { rPoint: rHex, s });
    expect(result.success).toBe(true);
  });
});

File: examples/zig/schnorr-zkp/SchnorrZKP_test.zig
const std = @import("std");
const root = @import("../examples_test.zig");
const runar = @import("runar");
const SchnorrZKP = @import("SchnorrZKP.runar.zig").SchnorrZKP;

fn decodeHexAlloc(allocator: std.mem.Allocator, hex: []const u8) ![]u8 {
    const out = try allocator.alloc(u8, hex.len / 2);
    _ = try std.fmt.hexToBytes(out, hex);
    return out;
}

fn contractPath(comptime basename: []const u8) []const u8 {
    return "schnorr-zkp/" ++ basename;
}

fn runCompileChecks(comptime basename: []const u8) !void {
    try root.runar.compileCheckSource(std.testing.allocator, @embedFile(basename), basename);
    try root.runar.compileCheckFile(std.testing.allocator, contractPath(basename));
}

test "compile-check SchnorrZKP.runar.zig" {
    try runCompileChecks("SchnorrZKP.runar.zig");
}

test "SchnorrZKP init stores pubKey" {
    const pub_key = runar.ecMulGen(11);
    const contract = SchnorrZKP.init(pub_key);
    try std.testing.expectEqual(runar.ecPointX(pub_key), runar.ecPointX(contract.pubKey));
    try std.testing.expectEqual(runar.ecPointY(pub_key), runar.ecPointY(contract.pubKey));
}

test "SchnorrZKP rejects invalid points" {
    try root.expectAssertFailure("schnorr-zkp-invalid-r-point");
}

test "SchnorrZKP verifies a valid Fiat-Shamir proof with wide bigint response" {
    const pub_key_hex =
        "fe8d1eb1bcb3432b1db5833ff5f2226d9cb5e65cee430558c18ed3a3c86ce1af" ++
        "07b158f244cd0de2134ac7c1d371cffbfae4db40801a2572e531c573cda9b5b4";
    const r_point_hex =
        "f01d6b9018ab421dd410404cb869072065522bf85734008f105cf385a023a80f" ++
        "0eba29d0f0c5408ed681984dc525982abefccd9f7ff01dd26da4999cf3f6a295";
    const s_le_hex =
        "eddbfe2cedf6f857ae5530a2dc2ee18f3f9562076f6269e09da736fee207ec5f";

    const pub_key = try decodeHexAlloc(std.testing.allocator, pub_key_hex);
    defer std.testing.allocator.free(pub_key);

    const r_point = try decodeHexAlloc(std.testing.allocator, r_point_hex);
    defer std.testing.allocator.free(r_point);

    const s_bytes = try decodeHexAlloc(std.testing.allocator, s_le_hex);
    defer std.testing.allocator.free(s_bytes);

    const contract = SchnorrZKP.init(pub_key);
    const s = runar.bin2num(s_bytes);

    contract.verify(r_point, s);
}
